tasks/main.yml

```diff
@@ -1,52 +1,132 @@
-  set_fact:
-    virt_install_cpu_model: "host-model"
-  when: ansible_distribution == "CentOS" and ansible_distribution_major_version == "7"
-  delegate_to: "{{ hyper }}"
+  - name: group hypervisors by distro
+    group_by:
+      key: "os_{{ ansible_distribution }}_{{ ansible_distribution_major_version }}"
+
+  - name: Check if VM exists already
+    virt:
+      name: "{{ inventory_hostname }}"
+      command: get_xml
+      uri: qemu:///system
+    register: guest_exists
+    ignore_errors: true
+    become: yes
+
+  - name: Ensure libvirt image directory is present
+    file:
+      path: "{{ image_path }}"
+      state: directory
+      mode: "0711"
+      owner: root
+      group: root
+    become: yes
+    when: guest_exists | failed
+
+  - name: create a remote temp directory
+    command: mktemp -d "{{ provision_tempdir }}/provision_vm.XXXX"
+    register: temp_dir
+    when: guest_exists | failed
+
+  - name: Set playbook temp_dir as fact
+    set_fact:
+      runtime_tempdir: "{{ temp_dir.stdout }}"
+    when: guest_exists | failed
+
+  - name: Change ownership of temp directory
+    file:
+      path: "{{ runtime_tempdir }}"
+      state: directory
+      mode: 0700
+      owner: qemu
+      group: qemu
+    become: yes
+    when: guest_exists | failed
 
-- name: copy over kickstart file
-  template: src="kickstart.cfg" dest="{{ kickstart_tempdir }}/{{ inventory_hostname }}.ks"
-  sudo: yes
   delegate_to: "{{ hyper }}"
+  # End of block
+
+- name: Prepare to start the guest using Kickstart
+  include: kickstart-guest.yml
+  when:
+    - guest_type is not defined or guest_type != "image"
+    - guest_exists | failed
 
-- name: copy over virt-install script
-  template: src="virt-install.sh" dest="{{ kickstart_tempdir }}/{{ inventory_hostname }}-install.sh"
-  sudo: yes
+- name: Prepare to start the guest using an OpenStack image
+  include: image-guest.yml
+  when:
+    - guest_type is defined
+    - guest_type == 'image'
+    - guest_exists | failed
+
+- name: Deploy script to deploy VM
+  template:
+    src: virt-install-script.sh.j2
+    dest: "{{runtime_tempdir}}/virt-install-script.sh.j2"
+  become: yes
   delegate_to: "{{ hyper }}"
+  when: guest_exists | failed
 
-- name: run virt-install script
-  command: "bash {{ kickstart_tempdir }}/{{ inventory_hostname }}-install.sh"
-  sudo: yes
-  environment: "{{ env_virt_install }}"
+- name: Start deployment of VM
+  command: "bash {{runtime_tempdir}}/virt-install-script.sh.j2"
   args:
     creates: "/etc/libvirt/qemu/{{ inventory_hostname }}.xml"
+  become: yes
+  environment: "{{ env_virt_install | default(omit) }}"
   delegate_to: "{{ hyper }}"
+  register: virt_install_job
+  when: guest_exists | failed
+
+  # Force ansible to know the IP address of this host so it can connect
+  # This fixes things if you don't have the IP address in DNS or /etc/hosts.
+- name: Set inventory_hostname to the new guest IP address
+  set_fact: inventory_hostname={{ internal_ip }}
```

Comment:
Redefining the inventory_hostname like this can have unintended consequences if later plays depend on the inventory_hostname being set to whatever is listed in the inventory. This is quite often the case in our plays. What I would suggest instead here is to use ansible_host (also known as ansible_ssh_host pre-2.0), which is a variable that defaults to inventory_hostname but can be used to tell Ansible what identity to use for the host for the purposes of SSH connections.

See: http://docs.ansible.com/ansible/intro_inventory.html#list-of-behavioral-inventory-parameters